File: src/main/scala/org/tindalos/principle/domain/agents/submodulesblueprint/SubmoduleDefinition.scala
package org.tindalos.principle.domain.agents.submodulesblueprint

import org.tindalos.principle.domain.core.PackageReference
import scala.collection.mutable.Set
import org.apache.commons.lang3.builder.HashCodeBuilder

import scala.collection.immutable.Seq

class SubmoduleDefinition(val id: SubmoduleId, val packages: scala.collection.immutable.Set[PackageReference]) {

  private val legalDependencies = Set[SubmoduleId]()

  def addPlannedDependencies(plannedDependencies: Seq[SubmoduleId]) = legalDependencies ++= plannedDependencies

  def overlapsWith(that: SubmoduleDefinition) = {
    val overlappingPackages = for (aPackage <- packages; otherPackage <- that.packages; if aPackage.oneContainsAnother(otherPackage))
    yield (aPackage, otherPackage)
    !overlappingPackages.isEmpty
  }

  def getLegalDependencies = legalDependencies.toSet

  override def equals(other: Any) =
    if (!other.isInstanceOf[SubmoduleDefinition]) false
    else other.asInstanceOf[SubmoduleDefinition].id.eq(id)

  override def hashCode() = new HashCodeBuilder().append(id).hashCode()

  override def toString() = "SubmoduleDefinition [" + id + "]"

}

File: src/main/java/org/tindalos/principle/domain/analyzers/submodulesblueprint/SubmoduleDefinition.java
package org.tindalos.principle.domain.analyzers.submodulesblueprint;

import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.tindalos.principle.domain.core.PackageReference;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Represents a module definition in a blueprint, containing the module's ID,
 * the packages it includes, and its planned dependencies on other modules.
 */
public class SubmoduleDefinition {

    private final SubmoduleId id;
    private final Set<PackageReference> packages;
    private final Set<SubmoduleId> legalDependencies;

    public SubmoduleDefinition(SubmoduleId id, Set<PackageReference> packages) {
        this.id = id;
        this.packages = Set.copyOf(packages); // Defensive copy, immutable
        this.legalDependencies = new HashSet<>(); // Mutable
    }

    public SubmoduleId id() {
        return id;
    }

    public Set<PackageReference> packages() {
        return packages;
    }

    public void addPlannedDependencies(List<SubmoduleId> plannedDependencies) {
        legalDependencies.addAll(plannedDependencies);
    }

    public boolean overlapsWith(SubmoduleDefinition that) {
        for (PackageReference aPackage : this.packages) {
            for (PackageReference otherPackage : that.packages) {
                if (aPackage.oneContainsAnother(otherPackage)) {
                    return true;
                }
            }
        }
        return false;
    }

    public Set<SubmoduleId> getLegalDependencies() {
        return Set.copyOf(legalDependencies);
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof SubmoduleDefinition)) {
            return false;
        }
        SubmoduleDefinition that = (SubmoduleDefinition) other;
        return this.id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(id).hashCode();
    }

    @Override
    public String toString() {
        return "SubmoduleDefinition [" + id + "]";
    }
}

